File: frontend/src/containers/lobby.js
import React, { Component } from 'react';
import { connect } from 'react-redux';
import Immutable from 'seamless-immutable';
import { Button } from 'rebass';
import PlayerList from '../components/playerList';

import { getPlayers } from '../redux/players';
import { getCurrentGame, actions } from '../redux/games';

class Lobby extends Component {
  getTitle() {
    if (this.props.currentGame) {
      return this.props.currentGame.name + ' — Lobby';
    } else {
      return 'What are you doing here? You haven\'t joined a game yet!';
    }
  }

  render() {
    return (
      <div>
        <h2>{this.getTitle()}</h2>
        <PlayerList players={this.props.players} />
        <Button onClick={this.props.startGame}>Start Game</Button>
      </div>
    );
  }
}

const mapStateToProps = state => {
  const game = getCurrentGame(state);
  return {
    currentGame: game,
    players: game ? getPlayers(state, game.players) : Immutable([]),
  };
};

const mapDispatchToProps = {
  startGame: actions.requestStartGame,
};

export default connect(mapStateToProps, mapDispatchToProps)(Lobby);

File: sw-ui-kt/src/main/kotlin/org/luxons/sevenwonders/ui/components/lobby/Lobby.kt
package org.luxons.sevenwonders.ui.components.lobby

import com.palantir.blueprintjs.Intent
import com.palantir.blueprintjs.bpButton
import org.luxons.sevenwonders.model.api.LobbyDTO
import org.luxons.sevenwonders.model.api.PlayerDTO
import org.luxons.sevenwonders.ui.redux.RequestStartGameAction
import org.luxons.sevenwonders.ui.redux.connectStateAndDispatch
import react.RBuilder
import react.RComponent
import react.RProps
import react.RState
import react.dom.*

interface LobbyStateProps : RProps {
    var currentGame: LobbyDTO?
    var currentPlayer: PlayerDTO?
}

interface LobbyDispatchProps : RProps {
    var startGame: () -> Unit
}

interface LobbyProps : LobbyDispatchProps, LobbyStateProps

class LobbyPresenter(props: LobbyProps) : RComponent<LobbyProps, RState>(props) {

    override fun RBuilder.render() {
        val currentGame = props.currentGame
        val currentPlayer = props.currentPlayer
        if (currentGame == null || currentPlayer == null) {
            div { +"Error: no current game." }
            return
        }
        div {
            h2 { +"${currentGame.name} — Lobby" }
            radialPlayerList(currentGame.players)
            if (currentPlayer.isGameOwner) {
                bpButton(
                    large = true,
                    intent = Intent.PRIMARY,
                    icon = "play",
                    title = currentGame.startAction.tooltip,
                    disabled = !currentGame.startAction.canDo,
                    onClick = { props.startGame() }
                ) {
                    + "START"
                }
            }
        }
    }
}

fun RBuilder.lobby() = lobby {}

private val lobby = connectStateAndDispatch<LobbyStateProps, LobbyDispatchProps, LobbyProps>(
    clazz = LobbyPresenter::class,
    mapStateToProps = { state, _ ->
        currentGame = state.currentLobby
        currentPlayer = state.currentPlayer
    },
    mapDispatchToProps = { dispatch, _ ->
        startGame = { dispatch(RequestStartGameAction()) }
    }
)
